test/com/google/javascript/jscomp/TypeInferenceTest.java

```diff
@@ -1516,6 +1516,40 @@ public void testNotIsArray() {
     verify("x", NUMBER_TYPE);
   }
 
+  public void testIsNumber() {
+    assuming("x", createUnionType(ARRAY_TYPE, NUMBER_TYPE));
+    inFunction("goog.asserts.assert(Number.isFinite(x));");
+    verify("x", NUMBER_TYPE);
+
+    assuming("x", createUnionType(ARRAY_TYPE, NUMBER_TYPE));
+    inFunction("goog.asserts.assert(Number.isInteger(x));");
+    verify("x", NUMBER_TYPE);
+
+    assuming("x", createUnionType(ARRAY_TYPE, NUMBER_TYPE));
+    inFunction("goog.asserts.assert(Number.isNaN(x));");
+    verify("x", NUMBER_TYPE);
+
+    assuming("x", createUnionType(ARRAY_TYPE, NUMBER_TYPE));
+    inFunction("goog.asserts.assert(Number.isSafeInteger(x));");
+    verify("x", NUMBER_TYPE);
+
+    assuming("x", createUnionType(ARRAY_TYPE, NUMBER_TYPE));
+    inFunction("goog.asserts.assert(!Number.isFinite(x));");
+    verify("x", NUMBER_TYPE);
+
+    assuming("x", createUnionType(ARRAY_TYPE, NUMBER_TYPE));
+    inFunction("goog.asserts.assert(!Number.isInteger(x));");
```

Comment:
I think the intention is that `x` should always be a number type: https://github.com/google/closure-compiler/blob/master/externs/es6.js#L1525-L1531.